feat:Added the periodically clearing serverlog function

XiaoLiang2333 opened the pull request:
#2811 
Configure the log-retention-time field in pika.conf. Serverlogs that exceed this time period are cleared

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
  - Introduced a configuration option for log retention time, enhancing log management capabilities.
  - Added methods for targeted log purging, allowing more precise control over server and binlog file retention.

- **Bug Fixes**
  - Improved server performance with detailed log purging logic based on specified retention periods.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

XiaoLiang2333 committed:
Added the serverlog periodic reclamation function

XiaoLiang2333 committed:
Revert "Added the serverlog periodic reclamation function"

This reverts commit d6262d7810ead886449cacba6a0a4f313ee61be6.

XiaoLiang2333 committed:
Added the serverlog periodic reclamation function

cheniujh commented on src/pika_server.cc:
```diff
@@ -1206,7 +1208,82 @@ void PikaServer::AutoCompactRange() {
+    if (interval_days > retention_time) {
+      std::string log_file = log_path + "/" + file;
+      LOG(INFO) << "Deleting out of date log file: " << log_file;
+      pstd::DeleteFile(log_file);
```
pls check return value and print Error Log if deleteFile failed

cheniujh commented on src/pika_server.cc:
```diff
@@ -1206,7 +1208,82 @@ void PikaServer::AutoCompactRange() {
+    int64_t now_timestamp = mktime(&now_time);
+    int64_t interval_days = (now_timestamp - log_timestamp) / 86400;
+
+    if (interval_days > retention_time) {
```
A logic is needed to guarantee that the latest log(which is in use) will not be deleted even if it's  logically expired. 
The time extracted from log file name is the time of it's creation. thus it's possible that a log has logically expired but still in use if the pika server is inactive.

XiaoLiang2333 replied:
Do your mean that assuming a Pika instance has been running continuously for 8 days, a server log created 8 days ago should logically have expired, but in reality, it is the most recent and currently used server log?

cheniujh replied:
yes, that's what i mean

XiaoLiang2333 committed:
Update AutoServerlogPurge()

Check return value and print Error Log if deleteFile failed

XiaoLiang2333 committed:
Refactor AutoServerlogPurge function

1. retain the newest serverlog file even if it has been expired
2. Make sure that serverlogs of [INFO],[ERROR], and [WARNING] levels are retained

cheniujh commented:
目前看着代码是OK的，能否辛苦实测一下呢，可以先手动mock一下代码，把过期时间改成10分钟，临时添加一些日志输出（比如每条指令都打一次日志，然后用benchmark压，很快就会累计大量日志，并且发生切换），主要测试一下功能正确性。

因为要上实际生产，所以需要稳妥一些，辛苦了！

cheniujh committed:
Merge branch 'unstable' into logchange

XiaoLiang2333 commented:
> 目前看着代码是OK的，能否辛苦实测一下呢，可以先手动mock一下代码，把过期时间改成10分钟，临时添加一些日志输出（比如每条指令都打一次日志，然后用benchmark压，很快就会累计大量日志，并且发生切换），主要测试一下功能正确性。
> 
> 因为要上实际生产，所以需要稳妥一些，辛苦了！

ok,I got it.I have done a simple test before, and I will do a more detailed test according to the procedure you described

XiaoLiang2333 committed:
Update pika_server.cc

Adjust the time calculation method

XiaoLiang2333 commented:
I ran more detailed tests,It seems to be working rightly.

Test approach:
1. Change the automatic clean-up time to 5 min.
2. Modify FLAGS_max_log_size of PikaGlogInit to 1(MB).
3. Add a lot of logging output to DispatchThread::HandleNewConn.
4. Run pika_benchmark

XiaoLiang2333 committed:
Optimize delete logic

XiaoLiang2333 committed:
update

XiaoLiang2333 committed:
Merge branch 'logchange' of https://github.com/XiaoLiang2333/pika into logchange
